### ISSUE TITLE
Control lifetime of task results

### ISSUE THREAD

**Issue by ebatz** [bloomen/transwarp#18]
The way I understand the current state of affairs is that all tasks (and their results) will live until the end of evaluating the whole task graph. For applications with sizable memory footprint, it would be nice to have a way to control the lifetime of results and mark intermediary objects that do not actually matter eventually as such, so that they can be cleaned up when they are not required by any further tasks.

Is there a mechanism to achieve this behavior?

**Comment by bloomen** [comment]
You can currently call reset() on a task which will invalidate the underlying shared future. If that’s the last reference then the memory is cleaned up. Are you basically looking for a mode where a task is auto-reset once every child has consumed the result?

**Comment by ebatz** [comment]
> Are you basically looking for a mode where a task is auto-reset once every child has consumed the result?

That's exactly right. Presumably there should be a mechanism to make this happen only if the user, when building the task graph, marks a task as intermediary, since the current behavior is that all results can be accessed later.

From a cursory look at the source code it seems like every task currently knows its parents, but parents do not know their children, correct? Broadly speaking, a way to implement the cleanup trigger would be a reference counter associated with a node that's incremented if a child registers itself as a consumer, and when a task finishes it'll decrement its parents' reference counters and call reset() on a parent if their reference counter is zero and they are marked as intermediary.

Does that sound like a viable way to go?

**Comment by bloomen** [comment]
Correct, only children know their parents.

We're currently working towards version 2.0 and expect to make breaking changes and already did (also switched to C++17). Now is a good time for a feature like this :)

I like the reference counting solution. But we need to obviously ensure that we only decrement when the child finished processing. Then we need a new method that lets the client opt-in to keep a task's result around. 

Are you interested in having a go at this?

**Comment by ebatz** [comment]
I realized that the reference-counter approach can be implemented with stuff already in transwarp using custom data (to store the refcount), a listener (to trigger the reset()) and some not-so-nice accessors.

Here's a short (and incomplete, because the check for parent ref count is omitted!) recipe,

```
#include <iostream>
#include "transwarp.h"

namespace tw = transwarp;

struct Data {
  unsigned int d;

  Data(unsigned int _d) : d(_d) { std::cout<<"Constructor "<<d<<std::endl; }
  Data(const Data& rhs) : d(rhs.d) { std::cout<<"Copy Constructor "<<d<<std::endl; }
  ~Data() { std::cout << "Destructor "<<d<<std::endl; }
  
};


class GarbageCollector : public tw::listener {
public:
    void handle_event(tw::event_type event, const std::shared_ptr<tw::node>& node) override { 
      if (event == tw::event_type::after_finished) {
        std::cout << "Finished "<<node->id()<<" ("
                  <<std::any_cast<const unsigned int&>(node->custom_data())
                  <<")"<<std::endl;
        for (auto aPar : node->parents())
          aPar.reset();
      }
    }
};


int main() {
    auto gc = std::make_shared<GarbageCollector>();

      // task1
    auto task1 = tw::make_task(tw::root, [](){ return Data(2); });
    task1->add_listener(gc);
    task1->set_custom_data(0u);

      // task2 consuming task1
    auto task2 = tw::make_task(tw::consume, [](const Data& dat){ return Data(dat.d+1); }, task1);
    task2->add_listener(gc);
    task2->set_custom_data(0u);
      // 'loop' over parents
    task1->set_custom_data(std::any_cast<unsigned int>(task1->node()->custom_data())+1);

      // task3 consuming task2
    auto task3 = tw::make_task(tw::consume, [](const Data& dat){ return Data(dat.d+1); }, task2);
    task3->add_listener(gc);
    task3->set_custom_data(0u);


      // get rid of our reference here
    task1.reset();
    task2.reset();

    std::cout << "Starting scheduling" <<std::endl;
    task3->schedule_all();
    std::cout << "result = " << task3->get().d << std::endl;
}
```

but the output is somewhat confusing to me:

```
Starting scheduling
Constructor 2
Copy Constructor 2
Destructor 2
Finished 0 (1)
Constructor 3
Copy Constructor 3
Destructor 3
Finished 1 (0)
Constructor 4
Copy Constructor 4
Destructor 4
Finished 2 (0)
result = 4
Destructor 2
Destructor 3
Destructor 4
```

I would have expected the final Destructors to get called as soon as the corresponding child task finishes and has the listener call reset(). Any clue about why that doesn't seem to be happening?

On a side note, the instrumented `Data` class suggests that there's an extra copy when returning from a task. Does that mean I should use (smart) pointers to pass data around in production code instead?

**Comment by bloomen** [comment]
`handle_event` should never allocate/deallocate memory as it may be some high-priority thread that's calling it. Neither can it lock a regular mutex, spin lock should be ok (like in tw::timer).

We cannot use `custom_data` as that should be solely reserved for user data.

In the `GarbageCollector` you call `reset` on a node rather than a task. A node in transwarp is a simple container for task metadata and every task has exactly one node (as a shared_ptr). Hence, the `Data` object is not destroyed at that point. You'd want to call `reset` on the task object or invalidate the future directly.

Before any scheduling, resetting a task has no effect as that simply invalidates the underlying future, but the future never became valid because we never scheduled. So the two resets after `// get rid of our reference here` have no effect.

Regarding your second question: Large objects whose memory should be shared by tasks should be passed around as pointers or references because getting a task result will always pass by reference. It has to because results are shared among children. 

You're getting the copy constructors called because your `Data` class is not movable (it defines a destructor -> rule of 5). If you define an explicit move constructor you will see that the object is moved and then the moved-from object is destroyed. 

I don't think we can use listeners to solve this as their finished event is called on the thread the task is run on, plus, we'd need to wait for ALL children to finish using the consumed data. What we want is for the objects to be destroyed on either the thread where `schedule` was called (main thread in your case) or some user defined thread accessible through an executor.

Does any of that make sense?

Ok, that's my two cents for now :) I'll have a think about this some more...

**Comment by ebatz** [comment]
> We cannot use `custom_data` as that should be solely reserved for user data.

Sorry, I should been been clearer on my intention to keep the proposed functionality in userland instead of adding a new feature to transwarp. In that sense, this **is** user data, no?

> In the `GarbageCollector` you call `reset` on a node rather than a task. A node in transwarp is a simple container for task metadata and every task has exactly one node (as a shared_ptr). Hence, the `Data` object is not destroyed at that point. You'd want to call `reset` on the task object or invalidate the future directly.

Thanks for the pointer (no pun intended). The event system only passes a `node` object in, which does not know about the task object it belongs to it seems like. That's fatal to the idea of (ab)using the event system, because from within the event system there does not seem to be a way to access the relevant task?!
Is there a reason to have tasks and nodes separated? You say that for each task there is one node; I suspect the opposite is also true, and if so, does it make sense to split 'content' and 'metadata' across two classes?

> Before any scheduling, resetting a task has no effect as that simply invalidates the underlying future, but the future never became valid because we never scheduled. So the two resets after // get rid of our reference here have no effect.

The intention was to clear the shared pointer, not to call `reset()` of the underlying task, hence `task1.reset()`, not `task1->reset()`. This was meant to make sure that there was no other reference as per your comment earlier

>  If that’s the last reference then the memory is cleaned up.

but since that comment referred to the future, not a reference to the task, those calls might indeed be superfluous.

> I don't think we can use listeners to solve this as their finished event is called on the thread the task is run on, plus, we'd need to wait for ALL children to finish using the consumed data.

The idea would be to decrement the ref count stored in custom_data. If the listener finds the decremented value equal to zero (i.e. this was the last task to consume the data), it would call reset() on the parent task. So the executing thread does not have to wait. The only issues I see are making the decrement of the ref count thread-safe, and the fact that the task/future is not accessible to the listener.

Does your comment about (de)allocation in listeners still apply in this scenario?


[Thanks for the clarification on the lack of a move constructor. Provided that move semantics is properly implemented, there shouldn't be a reason to pass smart pointers around then.]

**Comment by bloomen** [comment]
Yeah, there’s no way to get access to the task from a listener or an executor directly. You just get the node handed to you. The reason for this design choice is to prevent accidentally messing with a task object and doing unintended things with it (like rescheduling it or something). 

You’re right, task and node have a one-to-one relationship. Does it still make sense to have a node? I think so, for the above reasons. However, I could be convinced of otherwise. But even if we change this we still cannot call `reset` in `handle_event`.

Let me know if I follow your thoughts correctly:

We create a listener that listens to two events: `before_scheduled` (?) and `after_finished`.

The listener _has_ two objects: An executor and a map of (node, pair(ref_count, task)). (The `tasks()` method gives you pointers to all tasks)

In `before_scheduled` we increment the ref_count for the parent of the incoming node. 

In `after_finished` we decrement the ref_count for the parent of the incoming node. If the ref_count goes to zero we push a functor into the listener’s executor. The functor could look like: [task]{ task->reset(); } where `task` comes from a map lookup. By default, the executor could just be `tw::sequential` but can be overwritten by the user.

Does that seem reasonable?

A problem I see here is that `tasks()` returns raw pointers and not `weak_ptr`s. So when the executor comes around to running the functor and `task` may have vanished already, we have no way of detecting it. 

I think it would be pretty cool if we had a new listener class that the user can just instantiate (maybe configure it with a custom executor), then add it to the graph and get automatic garbage collection.

**Comment by ebatz** [comment]
> But even if we change this we still cannot call `reset` in `handle_event`.

Where does this restriction come in? Is it a design choice, or is this some fundamental requirement to do with thread safety or something?

The general flow looks good to me, just a couple of comments:
- To my taste the ref_count should be incremented as soon as the task graph is built. Otherwise one might end up in a situation where one final task is being scheduled and might finish while other tasks re-using intermediary results are still being scheduled, triggering cleanup too soon. This seems to be a matter of taste though: I build task graphs because I want to run all parts of it, but the transwarp philosophy seems to be that you only start caring about a part of a task graph as soon as it gets scheduled.
tl;dr: I'd prefer to increment the ref_count when creating tasks instead of at `before_scheduled`, but I might be missing something.

- At any rate, I think I'd be in favor of consolidating node and task into one. That should solve the invalid pointer problem, no? The listener is passed a standard shared pointer and (depending on why we can't call `reset()` from `handle_event`) can either schedule a cleanup task or do cleanup directly.

- Assuming that task and node will be consolidated, an alternative to the 'global' bookkeeping you describe would be to add a ref_count to each task/node individually. The advantage is that this avoids a large lookup in your map, the disadvantage being that we're increasing overhead on every task/node irrespective of whether the user actually wants garbage collection. Then again, there already are the timing numbers in a node that the same complaint applies to.

**Comment by bloomen** [comment]
> Where does this restriction come in? Is it a design choice, or is this some fundamental requirement to do with thread safety or something?

The issue is that `handle_event` is called on a thread that could potentially have strict time requirements (such as in audio processing). So heap (de)allocations or locking a mutex is not acceptable. So for our deallocator (not a fan of "garbage collector") if we can't do it directly we should at least make it configurable (through an executor) to deallocate on a specific, non-critical thread.

I agree, I think `ref_count` should be a field of a task. Does it need to be incremented on task creation? Not really, because when you call `schedule_all` all tasks will be scheduled synchronously so you will get the `before_scheduled` event for all tasks before any other events.

Ok, I am convinced to change `listener` and `executor` interfaces to deal with `itask` instead of `node`. `node` should be merged into `itask`. I created two separate issues for that:
https://github.com/bloomen/transwarp/issues/19
https://github.com/bloomen/transwarp/issues/20

I won't have much time in the next few weeks so you're more than welcome to take a stab at these two issues. This should give you a good opportunity to become an expert of the code base.

**Comment by bloomen** [comment]
@ebatz I'd be interested to know what you think of this: https://github.com/bloomen/transwarp/commit/aa0cc7c96711c61cd2e43f62e4573d88b67ab65a
Cheers!

**Comment by bloomen** [comment]
Done and dusted.
